stan/math/rev/functor/adj_jac_apply.hpp

```diff
@@ -19,53 +19,35 @@ namespace math {
 namespace internal {
 
 /**
- * Place the data for the y_adj_ into the appropriate container.
- */
-template <typename T, typename = void>
-struct build_y_adj;
-
-/**
- * Specialization returns the real adjoint value for arithmetic and eigen
- * types. Since `vari_value<Eigen>`` holds the dimensions we don't need M.
+ * Store the adjoint in y_vi[0] in y_adj
+ *
+ * @tparam size dimensionality of M
+ * @param[in] y_vi pointer to pointer to vari
+ * @param[in] M shape of y_adj
+ * @param[out] y_adj reference to variable where adjoint is to be stored
  */
 template <typename T>
-struct build_y_adj<T,
-                   require_t<disjunction<std::is_arithmetic<T>, is_eigen<T>>>> {
-  /**
-   * Retreive the adjoints for y_adj_
-   *
-   * @tparam size dimensionality of M
-   * @param[in] y_vi pointer to pointer to vari
-   * @param[in] M shape of y_adj (not used in this specialization)
-   */
-  template <typename VariType, size_t size>
-  static inline auto& apply(VariType&& y_vi, const std::array<int, size>& M) {
-    return y_vi->adj_;
-  }
-};
+inline auto build_y_adj(vari_base** y_vi, int M) {
+  return dynamic_cast<T>(y_vi[0])->adj_;
+}
 
 /**
- * Specialization returns the real adjoint value for std vector return types.
+ * Store the adjoints from y_vi in y_adj
+ *
+ * @tparam size dimensionality of M
+ * @param[in] y_vi pointer to pointers to varis
+ * @param[in] M shape of y_adj
+ * @param[out] y_adj reference to std::vector where adjoints are to be stored
  */
-template <typename T>
-struct build_y_adj<T, require_std_vector_t<T>> {
-  /**
-   * Store the adjoints from y_vi in y_adj
-   *
-   * @tparam size dimensionality of M
-   * @param[in] y_vi pointer to pointers to varis
-   * @param[in] M shape of y_adj
-   */
-  template <typename VariType, size_t size>
-  static inline T apply(VariType& y_vi, const std::array<int, size>& M) {
-    T y_adj;
-    y_adj.reserve(M[0]);
-    for (size_t m = 0; m < M[0]; ++m) {
-      y_adj.emplace_back(y_vi[m]->adj_);
-    }
-    return y_adj;
+template <>
+inline auto build_y_adj<vari**>(vari_base** y_vi, int M) {
+  std::vector<double> y_adj;
+  y_adj.reserve(M);
+  for (size_t m = 0; m < M; ++m) {
+    y_adj.emplace_back(dynamic_cast<vari*>(y_vi[m])->adj_);
```

Comment:
Why did you change this? idt this will work in general?

Reply:
The `adj_jac_apply` tests weren't working (some sort of out of bounds indexing) so I just desperately rewrote the varis logic to use fixed length tuples that hold `std::nullptr_t` s where non-autodiff types were.

This was the end of a long series of changes lol.

There are two hidden assumptions in this code regarding matrix types, but:

1. `adj_jac_apply` only deals on input with `var_value<MatrixXd>` matrices (they're converted into this format before they go in). This could be relaxed.

2. `adj_jac_apply` only deals with `var_value<MatrixXd>` matrices on the output. I don't think we should relax this.

3. `adj_jac_apply` only deals with arrays of scalars.

By assumption 2 & 3, if the output is of type `vari**` it is a `std::vector<var>`.

Reply:
> There are two hidden assumptions

Edit: At least two lol

Reply:
We should also relax assumption 3, eventually (I think 1 comes first though).